### DISCUSSION TITLE
Unit testing code that is tainted with Airflow imports

### DISCUSSION THREAD

**Discussion by p-hebert** [apache/airflow#18416] [Q&A]
**EDIT:** This whole thing is a red herring. See this [SO question](https://stackoverflow.com/questions/69276878/pytest-unittest-mock-patch-function-from-module) for an answer to the issue (unrelated to Airflow)

---

Hello!

I am faced with an issue that I have already invested a lot of hours in unfortunately - I would like to be able to mock large swaths of the airflow package for unit testing.

I want to be able to do DAG integrity tests in the same suite as python_callable unit tests, and I cannot figure out how to patch Airflow in a way that it does not break.

An example:

When attempting to use PostgresHook().get_records, I am faced with OperationalError:

https://gist.github.com/philippefutureboy/2c27a3727b7f71c53587ff4f965e26b8#file-operationalerror-log-L1

I have tried many things, including:

- session scoped autouse pytest fixture for PostgresHook:

`conftest.py`:

```py
@pytest.fixture(autouse=True, scope="session")
def mock_pg_hook():
    with mock.patch("airflow.hooks.postgres_hook.PostgresHook") as MockPostgresHook:
        yield MockPostgresHook.return_value
```

- session scoped autouse pytest fixture mocking get_connection (inspired from [here](https://godatadriven.com/blog/testing-and-debugging-apache-airflow/)):

```py
@pytest.fixture(autouse=True, scope="session")
def mock_get_conn():
    from airflow.hooks.base_hook import BaseHook
    # from airflow.models import Connection
    mock_conn = mock.MagicMock()
    with mock.patch.object(
        BaseHook,
        "get_connection",
        mock_conn
        # alternatively, a Connection could be created, but that means I need a backend
        # return_value=Connection(schema="https", host="api.sunrise-sunset.org"),
    ):
        yield
```

which results in the following error (we are on GCP and we don't have the amazon extras, I don't see why this is an issue):

```
self = <airflow.providers.postgres.hooks.postgres.PostgresHook object at 0x2aae82220>
conn = <MagicMock name='mock()' id='11457583184'>

    def get_iam_token(self, conn: Connection) -> Tuple[str, str, int]:
        """
        Uses AWSHook to retrieve a temporary password to connect to Postgres
        or Redshift. Port is required. If none is provided, default is used for
        each service
        """
>       from airflow.providers.amazon.aws.hooks.base_aws import AwsBaseHook
E       ModuleNotFoundError: No module named 'airflow.providers.amazon'

/Users/philippe/Library/Caches/pypoetry/virtualenvs/pinkdolphin-pipeline-7zGJZnY7-py3.8/lib/python3.8/site-packages/airflow/providers/postgres/hooks/postgres.py:171: ModuleNotFoundError
```

- file-scoped mock.patch around the affected tests:

```py
with mock.patch("airflow.hooks.postgres_hook.PostgresHook") as MockPostgresHook:
    # import file under test
    # write tests
```

- tried to setup a sqlite3 db for the tests, as is done in [apache/airflow:tests/conftest.py](https://github.com/apache/airflow/blob/main/tests/conftest.py).

```py
@pytest.fixture(autouse=True, scope="session")
def setup(testrun_uid): # from pytest-xdist
    airflow_conn_uri = f'/tmp/airflow-{testrun_uid}.db'
    open(airflow_conn_uri, 'a').close()
    os.environ['AIRFLOW__CORE__SQL_ALCHEMY_CONN'] = airflow_conn_uri

    from airflow.utils import db

    db.resetdb()
    yield
```

This leads to another error where I need to set a connection, which is not what I want - it slows down tests and make them more fragile.

---

So as you can see, I'm starting to be out of options when it comes to mocking Airflow. I'm starting to strongly consider either completely mock the airflow package from the root or refactor and extract all references to Airflow in any non-DAG file. Until I have an answer I'll probably focus on the second option 🤷‍♂️ 

Has anyone had any success with mocking airflow?

**Comment by iostreamdoth** [comment]
I m failing to understand this. Could you explain what do you mean by DAG integrity test? Is it like you want to verify if these tasks are in expected order? Or are you writing your own operators?

**Comment by p-hebert** [comment]
Yep! DAG integrity tests is just a fancy term for "does this DAG compile or does it have errors?"
I've also updated the gist with a test file for the dag integrity tests, with a link to the original book proposing the technique in the docstring of the module.

https://gist.github.com/philippefutureboy/2c27a3727b7f71c53587ff4f965e26b8

And to avoid confusion - DAG integrity tests are not the same tests as the domain logic tests that I try to run in the same test suite, which are the ones that are failing.

**Reply by iostreamdoth** [reply]
https://github.com/apache/airflow/blob/main/tests/always/test_example_dags.py I've stumbled across this could be helpful.

**Reply by p-hebert** [reply]
Thanks for the suggestion :)
I don't think that in this case it solves the problem - it seems my issue is really located to the import of PostgresHook / Variable in my src files. As a simple test I commented out the code in the dag integrity test files, and I still get the error.

**Comment by p-hebert** [comment]
Opened this thread on SO (same issue happens after refactoring of imports to be done in a function). Most likely this is related to my handling of pytest/unittest:

https://stackoverflow.com/questions/69276878/pytest-unittest-mock-patch-function-from-module

**Comment by p-hebert** [comment]
@potiuk Thanks for your answer on the SO thread - this was entirely unrelated to Airflow, and much more related to my usage of unittest.mock.patch!
